File: springboot-jwt/src/main/java/com/grepp/spring/infra/auth/token/JwtAuthenticationFilter.java
package com.grepp.spring.infra.auth.token;

import com.grepp.spring.app.model.auth.token.RefreshTokenRepository;
import com.grepp.spring.app.model.auth.token.UserBlackListRepository;
import com.grepp.spring.app.model.auth.token.entity.RefreshToken;
import com.grepp.spring.app.model.auth.token.entity.UserBlackList;
import com.grepp.spring.infra.error.exceptions.CommonException;
import com.grepp.spring.infra.response.ResponseCode;
import io.jsonwebtoken.Claims;
import io.jsonwebtoken.ExpiredJwtException;
import jakarta.servlet.FilterChain;
import jakarta.servlet.ServletException;
import jakarta.servlet.http.Cookie;
import jakarta.servlet.http.HttpServletRequest;
import jakarta.servlet.http.HttpServletResponse;
import java.io.IOException;
import java.util.Arrays;
import java.util.Optional;
import java.util.UUID;
import lombok.RequiredArgsConstructor;
import org.springframework.http.ResponseCookie;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.stereotype.Component;
import org.springframework.web.filter.OncePerRequestFilter;

@Component
@RequiredArgsConstructor
public class JwtAuthenticationFilter extends OncePerRequestFilter {

    private final JwtProvider jwtProvider;
    private final RefreshTokenRepository refreshTokenRepository;
    private final UserBlackListRepository userBlackListRepository;

    @Override
    protected void doFilterInternal(HttpServletRequest request, HttpServletResponse response,
        FilterChain filterChain) throws ServletException, IOException {

        String requestAccessToken = resolveToken(request, TokenType.ACCESS_TOKEN);
        if (requestAccessToken == null) {
            filterChain.doFilter(request, response);
            return;
        }

        Claims claims = jwtProvider.parseClaim(requestAccessToken);

        if(request.getRequestURI().equals("/auth/logout")){
            Optional<RefreshToken> optional = refreshTokenRepository.findByAccessTokenId(claims.getId());
            optional.ifPresent(refreshTokenRepository::delete);
            filterChain.doFilter(request, response);
            return;
        }

        if(userBlackListRepository.existsById(claims.getSubject())){
            filterChain.doFilter(request, response);
            return;
        }

        try {
            if(jwtProvider.validateToken(requestAccessToken)){
                Authentication authentication = jwtProvider.genreateAuthentication(requestAccessToken);
                SecurityContextHolder.getContext().setAuthentication(authentication);
            }
        } catch (ExpiredJwtException ex) {
            String newAccessToken = renewingAccessToken(requestAccessToken, request);
            if (newAccessToken == null) {
                filterChain.doFilter(request, response);
                return;
            }
            RefreshToken newRefreshToken = renewingRefreshToken(claims.getId());
            responseToken(response, newAccessToken, newRefreshToken);
        }

        filterChain.doFilter(request, response);
    }

    private void responseToken(HttpServletResponse response, String newAccessToken,
        RefreshToken newRefreshToken) {

        ResponseCookie accessTokenCookie =
            TokenCookieFactory.create(TokenType.ACCESS_TOKEN.name(), newAccessToken,
                3000000L);

        ResponseCookie refreshTokenCookie =
            TokenCookieFactory.create(TokenType.REFRESH_TOKEN.name(), newRefreshToken.getToken(),
                jwtProvider.getAtExpiration());

        response.addHeader("Set-Cookie", accessTokenCookie.toString());
        response.addHeader("Set-Cookie", refreshTokenCookie.toString());
    }

    private RefreshToken renewingRefreshToken(String id) {
        RefreshToken refreshToken = refreshTokenRepository.findByAccessTokenId(id).get();
        refreshToken.setToken(UUID.randomUUID().toString());
        refreshTokenRepository.save(refreshToken);
        return refreshToken;
    }

    private String renewingAccessToken(String requestAccessToken, HttpServletRequest request) {
        Authentication authentication = jwtProvider.genreateAuthentication(requestAccessToken);
        String refreshToken = resolveToken(request, TokenType.REFRESH_TOKEN);
        Claims claims = jwtProvider.parseClaim(requestAccessToken);
        Optional<RefreshToken> optional = refreshTokenRepository.findByAccessTokenId(claims.getId());
        if (optional.isEmpty()) {
            return null;
        }

        RefreshToken storedRefreshToken = optional.get();
        if (!storedRefreshToken.getToken().equals(refreshToken)) {
            userBlackListRepository.save(new UserBlackList(authentication.getName()));
            throw new CommonException(ResponseCode.SECURITY_INCIDENT);
        }

        return generateAccessToken(authentication);
    }

    private String generateAccessToken(Authentication authentication) {
        String newAccessToken = jwtProvider.generateAccessToken(authentication).getToken();
        SecurityContextHolder.getContext().setAuthentication(authentication);
        return newAccessToken;
    }

    private String resolveToken(HttpServletRequest request, TokenType tokenType) {
        String headerToken = request.getHeader("Authorization");
        if (headerToken != null && headerToken.startsWith("Bearer")) {
            return headerToken.substring(7);
        }

        Cookie[] cookies = request.getCookies();
        if (cookies == null) {
            return null;
        }

        return Arrays.stream(cookies)
            .filter(e -> e.getName().equals(tokenType.name()))
            .map(Cookie::getValue).findFirst()
            .orElse(null);

    }
}

File: grepp-gateway/src/main/kotlin/com/grepp/spring/token/filter/JwtAuthenticationFilter.kt
package com.grepp.spring.token.filter

import com.grepp.infra.response.ResponseCode
import com.grepp.spring.infra.error.exceptions.CommonException
import com.grepp.spring.token.code.GrantType
import com.grepp.spring.token.code.TokenType
import com.grepp.spring.token.dto.AccessTokenDto
import com.grepp.spring.token.dto.TokenDto
import com.grepp.spring.token.entity.RefreshToken
import com.grepp.spring.token.entity.UserBlackList
import com.grepp.spring.token.repository.UserBlackListRepository
import com.grepp.spring.token.service.RefreshTokenService
import com.grepp.spring.token.util.JwtProvider
import com.grepp.spring.token.util.TokenResponseExecutor
import io.jsonwebtoken.Claims
import io.jsonwebtoken.ExpiredJwtException
import jakarta.servlet.FilterChain
import jakarta.servlet.ServletException
import jakarta.servlet.http.HttpServletRequest
import jakarta.servlet.http.HttpServletResponse
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Component
import org.springframework.web.filter.OncePerRequestFilter
import java.io.IOException

@Component
class JwtAuthenticationFilter(
    val jwtProvider: JwtProvider,
    val refreshTokenService: RefreshTokenService,
    val userBlackListRepository: UserBlackListRepository
) : OncePerRequestFilter() {

    private val log = LoggerFactory.getLogger(javaClass)

    @Throws(ServletException::class)
    protected override fun shouldNotFilter(request: HttpServletRequest): Boolean {
        val excludePath: MutableList<String> = ArrayList()
        excludePath.addAll(listOf("/error", "/favicon.ico", "/css", "/img", "/js", "/download"))
        excludePath.addAll(listOf("/member/signup", "/member/signin"))
        val path = request.requestURI
        return excludePath.stream().anyMatch { prefix: String? ->
            path.startsWith(
                prefix!!
            )
        }
    }

    @Throws(ServletException::class, IOException::class)
    protected override fun doFilterInternal(
        request: HttpServletRequest,
        response: HttpServletResponse,
        filterChain: FilterChain
    ) {
        log.info(request.requestURI)

        val requestAccessToken = jwtProvider.resolveToken(request, TokenType.ACCESS_TOKEN)

        if(requestAccessToken == null) {
            filterChain.doFilter(request, response)
            return
        }

        val claims: Claims = jwtProvider.parseClaim(requestAccessToken)
        if (userBlackListRepository.existsById(claims.subject)) {
            filterChain.doFilter(request, response)
            return
        }

        try {
            if (jwtProvider.validateToken(requestAccessToken)) {
                request.setAttribute("x-member-id", claims.subject)
                // fixme accesstoken 발행 시 권한 추가
                request.setAttribute("x-member-role", claims["roles"])
            }
        } catch (ex: ExpiredJwtException) {
            val newAccessToken: AccessTokenDto? = renewingAccessToken(requestAccessToken, request)
            newAccessToken ?: return

            val newRefreshToken = renewingRefreshToken(claims.id, newAccessToken.id)
            request.setAttribute("x-member-id", claims.subject)
            request.setAttribute("x-member-role", claims["roles"])
            responseToken(response, newAccessToken, newRefreshToken)
        } finally {
            filterChain.doFilter(request, response)
        }
    }

    private fun responseToken(
        response: HttpServletResponse, newAccessToken: AccessTokenDto,
        newRefreshToken: RefreshToken
    ) {
        val tokenDto = TokenDto(
            accessToken = newAccessToken.token,
            refreshToken = newRefreshToken.token,
            grantType = GrantType.BEARER,
            atExpiresIn = jwtProvider.atExpiration,
            rtExpiresIn = jwtProvider.rtExpiration
        )

        TokenResponseExecutor.response(response, tokenDto)
    }

    private fun renewingRefreshToken(id: String, newTokenId: String): RefreshToken {
        return refreshTokenService.renewingToken(id, newTokenId)!!
    }

    private fun renewingAccessToken(
        requestAccessToken: String,
        request: HttpServletRequest
    ): AccessTokenDto? {
        val refreshToken: String? = jwtProvider.resolveToken(request, TokenType.REFRESH_TOKEN)

        val claims: Claims = jwtProvider.parseClaim(requestAccessToken)

        val storedRefreshToken: RefreshToken =
            refreshTokenService.findByAccessTokenId(claims.id)
                ?: return null

        if (storedRefreshToken.token != refreshToken) {
            userBlackListRepository.save(UserBlackList(claims.subject))
            throw CommonException(ResponseCode.SECURITY_INCIDENT)
        }

        return jwtProvider.generateAccessToken(claims.subject, claims["roles"].toString())
    }
}